Sources/Workspace/ToolsVersionSpecificationPrepender.swift

```diff
@@ -0,0 +1,92 @@
+///
+/// A `FileSystemError` is thrown if the manifest file is unable to be read from or written to.
+///
+/// - Precondition: `manifestDirectoryPath` must be a valid path to a directory that contains a `Package.swift` file.
+///
+/// - Parameters:
+///   - manifestDirectoryPath: The absolute path to the given directory.
+///   - toolsVersion: The Swift tools version to specify as the lowest supported version.
+///   - fileSystem: The filesystem to read/write the manifest file on.
+///
+/// - Throws: A `FileSystemError` instance, if the manifest file is unable to be located, read from, or written to..
+public func prependToolsVersionSpecification(toDefaultManifestIn manifestDirectoryPath: AbsolutePath, specifying toolsVersion: ToolsVersion, fileSystem: FileSystem) throws {
+    let manifestFilePath = manifestDirectoryPath.appending(component: Manifest.filename)
+    try prependToolsVersionSpecification(toManifestAt: manifestFilePath, specifying: toolsVersion, fileSystem: fileSystem)
+}
+
+// FIXME: Throw an error if the specified version is greater than the version-specific manifest's version?
+// For example, if the manifest file is Package@swift-4.0.swift and the given version to specify is 5.0.
+/// Prepends a Swift tools version specification to the specified manifest file.
+///
+/// If the main manifest file already contains a valid tools version specification (ignoring the validity of the version specifier and that of everything following it), then the existing specification is replaced by this new one.
+///
+/// The version specifier in the specification does not contain any build metadata or pre-release identifier. The patch version is included if and only if it's not zero.
+///
+/// A `FileSystemError` is thrown if the manifest file is unable to be located, read from, or written to.
+///
+/// - Precondition: `manifestFilePath` must be a valid path to a file.
+///
+/// - Parameters:
+///   - manifestFilePath: The absolute path to the specified manifest file.
+///   - toolsVersion: The Swift tools version to specify as the lowest supported version.
+///   - fileSystem: The filesystem to read/write the manifest file on.
+///
+/// - Throws: A `FileSystemError` instance, if the manifest file is unable to be read from or written to..
+public func prependToolsVersionSpecification(toManifestAt manifestFilePath: AbsolutePath, specifying toolsVersion: ToolsVersion, fileSystem: FileSystem) throws {
+    // FIXME: Throw a `FileSystemError` instead?
+    // The only problem is that there doesn't seem to be a `FileSystemError.Kind` case that describes this kind of error.
+    // Or, we can revert it back to an assert, and let `fileSystem.readFileContents(manifestFilePath)` throw an error if the file can't be found.
+    precondition(fileSystem.isFile(manifestFilePath), "cannot locate the manifest file at \(manifestFilePath)")
```

Comment:
I would like to draw attention to this FIXME. 

I changed this from an `assert` to a `precondition`. So now SwiftPM could potentially crash if the manifest file is not found at the path given. It's probably better if an error is thrown instead, but I can't find the correct error to throw in `TSCBasic.FileSystemError`. 

I could also change it back to an `assert`, which is ignored when SwiftPM is compiled for release, and the call to `FileSystem.readFileContents` 2 lines later throws an error if the manifest can't be found. The problem with this is that the error thrown (`FileSystemError.Kind.ioError` or `.noEntry` or `.isDirectory`) doesn't reflect the actual error.

Reply:
I'll take a look — this should definitely be an Error a not an assertion or precondition.

Reply:
I'll look at the existing code and the changes in the next few days and come up with a recommendation.

Reply:
I think that in this case, `noEntry` is probably the right kind of FileSystemError.  It should definitely not be an assertion or precondition.  The caller should be able to catch this error and report it.

I agree that this is not very specific, so the other, and maybe better, thing to do would be to define a specialization of Swift.Error (called for example ToolsVersionPresented.Error) similar to ToolsVersionLoader.Error that represents the semantic problem (no manifest) and then wraps the underlying error.

Reply:
I think I'll go with the latter option, so the user can receive a more specific feedback. Thanks for the suggestion!